### PR TITLE
User Dashboard - GitHub Repos

### LINKED ISSUE #14: User Dashboard: Github Repositories
**Details:** A user exists in the database. They have a Github token stored. Generate a token here for testing and development: https://github.com/settings/tokens

A future story will implement generating these tokens dynamically but that isn't necessary here.

```
As a logged in user
When I visit /dashboard
Then I should see a section for "Github"
And under that section I should see a list of 5 repositories with the name of each Repo linking to the repo on Github
```

- [x] API call uses: https://developer.github.com/v3/repos/#list-your-repositories
- [x] API call **DOES NOT** use: https://developer.github.com/v3/repos/#list-user-repositories
- [x] No hashes in the view.
- [x] Uses objects in the view to represent the Repos.
- [x] Edge case: Make sure this shows the proper repositories when there are more than one user in the database with different tokens.
- [x] Don't display the "Github" section if the user is missing a Github token

### PR DESCRIPTION
**Pull request by javier-aguilar**
- User can see 5 of their repositories listed on the dashboard (when provided with a token)

Fixes #14

### CODE REVIEW

**Comment by javier-aguilar** [comment]
Before merging, I'd like to integrate your testing @tony-gomes 
Also, check if there is anything that needs to be changed. 
Additionally, as discussed, the token column is not currently encrypted. This can/will most likely be changed in the future.

File: .ruby-version [hunk 1]
```diff
@@ -0,0 +1 @@
+2.4.1
```
Comment (tony-gomes) [inline .ruby-version:1]: I am running 2.5.3 based on the mod intermission guidelines... were we supposed to install 2.4.1 for this project?
https://backend.turing.io/module3/misc/ruby_and_rails_versions

File: app/models/github.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
+class Github
+  class Repo
```
Comment (tony-gomes) [inline app/models/github.rb:2]: Really like this as a module & follows the source syntax... great work!

File: app/models/user.rb [hunk 3]
```diff
@@ -7,4 +7,8 @@ class User < ApplicationRecord
   validates :first_name, presence: true
   enum role: { default: 0, admin: 1 }
   has_secure_password
+
+  def repositories
+    Github.list_repos(token)
```
Comment (tony-gomes) [inline app/models/user.rb:12]: I am not sure if this should live here... we may want to use a method call to this being created elsewhere? It just doesn't feel like a User should know how to create a Github...

File: app/views/users/show.html.erb [hunk 4]
```diff
@@ -10,4 +10,14 @@
   <section>
     <h1>Bookmarked Segments</h1>
   </section>
+  <% if current_user.token %>
```
Comment (tony-gomes) [inline app/views/users/show.html.erb:13]: This is great work using the guard clause!

File: config/routes.rb [hunk 5]
```diff
@@ -35,6 +35,10 @@
   # Is this being used?
   get '/video', to: 'video#show'
 
+  namespace :user do
```
Comment (tony-gomes) [inline config/routes.rb:38]: We might want to think about a GitHub specific route & controller now that it is starting to take shape as an object?

File: db/migrate/20200505224554_add_token_to_users.rb [hunk 6]
```diff
@@ -0,0 +1,5 @@
+class AddTokenToUsers < ActiveRecord::Migration[5.2]
+  def change
+    add_column :users, :token, :string
```
Comment (tony-gomes) [inline db/migrate/20200505224554_add_token_to_users.rb:3]: I wonder if there are any encryption considerations for this as we discussed & I think it is worth mentioning for guidance at our check-in...

File: db/migrate/20200505224554_add_token_to_users.rb [hunk 7]
```diff
@@ -0,0 +1,5 @@
+class AddTokenToUsers < ActiveRecord::Migration[5.2]
```
Comment (tony-gomes) [inline db/migrate/20200505224554_add_token_to_users.rb:1]: I like the idea of storing this & way to intuit it this early on!

File: spec/features/user/user_dashboard_spec.rb [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    @user2 = create(:user)
+    visit '/'
+    click_on "Sign In"
+    fill_in 'session[email]', with: @user.email
+    fill_in 'session[password]', with: @user.password
+    click_on 'Log In'
+  end
+  it "once logged in, can see their repositories if they have a token" do
+    @user.update(token: ENV['GITHUB_TOKEN_1'])
+    @user2.update(token: ENV['GITHUB_TOKEN_2'])
```
Comment (tony-gomes) [inline spec/features/user/user_dashboard_spec.rb:15]: Great use of figaro here!

**Review by tony-gomes — APPROVED** [review]
Great work Javi! I really dig the way you are thinking ahead & noticing things that we should consider.

We should discuss whether or not a facade is necessary to follow the origin & using the you_tube syntax to store that repo module...

File: .ruby-version (hunk 1 above)
Comment (javier-aguilar) [reply .ruby-version:1]: Yeah, that's what I had at first. But on this project readme, it states Ruby 2.4.1 so I went with that.

File: app/models/github.rb (hunk 2 above)
Comment (javier-aguilar) [reply app/models/github.rb:2]: Thanks 👍

File: app/models/user.rb (hunk 3 above)
Comment (javier-aguilar) [reply app/models/user.rb:12]: So a Github isn't being created here, the user is just calling the Github class which has a method to list a user's repos based on their token.

File: config/routes.rb (hunk 5 above)
Comment (javier-aguilar) [reply config/routes.rb:38]: For sure. Let's see what these next stories ask.

File: db/migrate/20200505224554_add_token_to_users.rb (hunk 6 above)
Comment (javier-aguilar) [reply db/migrate/20200505224554_add_token_to_users.rb:3]: Definitely!

**Pull request merged**